pkg/runner/job_status.go

```diff
@@ -0,0 +1,194 @@
+		}
+	}
+
+	testStepStatus.Events = filteredTestEvents
+	targetStatuses, err := jr.buildTargetStatuses(coordinates)
+	if err != nil {
+		return nil, fmt.Errorf("could not build target status for test step %s: %v", coordinates.TestStepLabel, err)
+	}
+	testStepStatus.TargetStatuses = targetStatuses
+	return &testStepStatus, nil
+}
+
+// buildTestStatus builds the status of a test belonging to a specific to a test
+func (jr *JobRunner) buildTestStatus(coordinates TestCoordinates, currentJob *job.Job) (*job.TestStatus, error) {
+
+	var currentTest *test.Test
+	// Identify the test within the Job for which we are asking to calculate the status
+	for _, candidateTest := range currentJob.Tests {
+		if candidateTest.Name == coordinates.TestName {
+			currentTest = candidateTest
+			break
+		}
+	}
+
+	if currentTest == nil {
+		return nil, fmt.Errorf("job with id %d does not include any test named %s", coordinates.JobID, coordinates.TestName)
+	}
+	testStatus := job.TestStatus{
+		TestName:         coordinates.TestName,
+		TestStepStatuses: make([]job.TestStepStatus, len(currentTest.TestStepsBundles)),
+	}
+
+	// Build a TestStepStatus object for each TestStep
+	for index, bundle := range currentTest.TestStepsBundles {
+		testStepCoordinates := TestStepCoordinates{
+			TestCoordinates: coordinates,
+			TestStepName:    bundle.TestStep.Name(),
+			TestStepLabel:   bundle.TestStepLabel,
+		}
+		status, err := jr.buildTestStepStatus(testStepCoordinates)
+		if err != nil {
+			return nil, fmt.Errorf("could not build TestStatus for test %s: %v", bundle.TestStep.Name(), err)
+		}
+		testStatus.TestStepStatuses[index] = *status
+	}
+	return &testStatus, nil
+}
+
+// BuildRunStatus builds the status of a run with a job
+func (jr *JobRunner) BuildRunStatus(coordinates RunCoordinates, currentJob *job.Job) (*job.RunStatus, error) {
+
+	runStatus := job.RunStatus{RunID: coordinates.RunID, TestStatuses: make([]job.TestStatus, len(currentJob.Tests))}
+
+	for index, currentTest := range currentJob.Tests {
+		testCoordinates := TestCoordinates{RunCoordinates: coordinates, TestName: currentTest.Name}
+		testStatus, err := jr.buildTestStatus(testCoordinates, currentJob)
+		if err != nil {
+			return nil, fmt.Errorf("could not rebuild status for test %s: %v", currentTest.Name, err)
+		}
+		runStatus.TestStatuses[index] = *testStatus
+	}
+	return &runStatus, nil
+}
+
+// BuildStatus builds the status of all runs belonging to the job
+func (jr *JobRunner) BuildRunStatuses(currentJob *job.Job) ([]job.RunStatus, error) {
+
+	runStatuses := make([]job.RunStatus, 0, currentJob.Runs)
+	for runID := uint(0); runID < currentJob.Runs; runID++ {
+		runCoordinates := RunCoordinates{JobID: currentJob.ID, RunID: types.RunID(runID)}
+		runStatus, err := jr.BuildRunStatus(runCoordinates, currentJob)
+		if err != nil {
+			return nil, fmt.Errorf("could not rebuild run status for run %d: %v", runID, err)
+		}
+		runStatuses[runID] = *runStatus
```

Comment:
this seems to fail for me on a 3-runs job:
```
panic: runtime error: index out of range [0] with length 0
```

Just change line 184:
```
runStatuses := make([]job.RunStatus, 0, currentJob.Runs)
```
to
```
runStatuses := make([]job.RunStatus, currentJob.Runs)
```
so it creates a slice with `Runs` elements

Reply:
Indeed, thanks, I barely tested this last night. I think I'll follow the same approach as the other control paths and `append` instead. Performance-wise should be the same?

Reply:
Fixed.